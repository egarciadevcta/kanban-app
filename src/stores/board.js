import { defineStore } from "pinia";

export const useBoardsStore = defineStore({
  id: "boards",
  state: () => ({
    boards: [],
    selectedBoard: 0,
    selectedColumn: 0,
    selectedCardTask: 0,
    draggedTask: 0,
    searchTerm: "",
    draggedCardTask: {},
  }),
  getters: {
    getSearchTerm: (state) => state.searchTerm,
    getColumns: (state) => state.boards[state.selectedBoard]?.columns,
    getCurrentBoard: (state) => state.boards[state.selectedBoard],
    getCurrentColumn: (state) =>
      state.boards[state.selectedBoard]?.columns[state.selectedColumn],
    getCardTask: (state) =>
      state.boards[state.selectedBoard]?.columns[state.selectedColumn]?.tasks[
        state.selectedCardTask
      ],
    getColumnsNames: (state) =>
      state.boards[state.selectedBoard]?.columns.map((c) => c.name),
    filteredTasks: (state) => (tasks) => {
      const result = tasks.filter((task) =>
        task.title.toLowerCase().includes(state.searchTerm.toLowerCase())
      );
      return result;
    },
    getDroppedTaskCard: (state) => state.draggedCardTask,
  },
  actions: {
    changeTaskColumn(index) {
      if (!(index === this.selectedColumn)) {
        this.getCurrentBoard?.columns[index]?.tasks.push(this.getCardTask);
        this.getCurrentColumn?.tasks.splice(this.selectedCardTask, 1);
        this.selectedColumn = index;
        this.selectedCardTask = this.getCurrentColumn?.tasks.length - 1;
      }
    },
    setSearchTerm(term) {
      this.searchTerm = term;
    },
    setDraggedCardTask(taskData) {
      this.draggedCardTask = taskData;
    },
    drop(newColumnIndex) {
      if (
        this.draggedCardTask &&
        newColumnIndex !== this.draggedCardTask.columnIndex
      ) {
        const { task, columnIndex, taskIndex } = this.draggedCardTask;
        // Mover la tarea a la nueva columna
        this.getCurrentBoard.columns[newColumnIndex]?.tasks.push(task);
        // Eliminar la tarea de la columna original
        this.getCurrentBoard.columns[columnIndex]?.tasks.splice(taskIndex, 1);
      }
    },
  },
});
