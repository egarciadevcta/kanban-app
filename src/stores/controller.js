import { defineStore } from "pinia";

export const useControllerStore = defineStore({
  id: "controller",
  state: () => ({
    cardTaskView: false,
    cardTaskForm: {
      visible: false,
      edit: false,
    },
    boardForm: {
      visible: false,
      edit: false,
    },
    delete: {
      visible: false,
      board: false,
    },
    overlay: false,
    dragging: false,
  }),
  getters: {},
  actions: {
    hideOverlay() {
      this.overlay = false;
      this.cardTaskView = false;
      this.cardTaskForm.visible = false;
      this.boardForm.visible = false;
      this.delete.visible = false;
    },
  },
});
