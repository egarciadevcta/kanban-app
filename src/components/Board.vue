<script setup>
import { Container, Draggable } from 'vue-dndrop';
import CardTask from "@/components/CardTask.vue";
import { useBoardsStore } from '@/stores/board.js';
import { useControllerStore } from '@/stores/controller.js';
import { computed, ref } from "vue";
import { bulletColors } from "../utils/helpers";

const controllerStore = useControllerStore();
const boardsStore = useBoardsStore();
const searchTerm = computed(() => boardsStore.getSearchTerm);
const columns = computed(() => boardsStore.getColumns);
const droppedTaskCard = computed(()=> boardsStore.getDroppedTaskCard);

const onClickTask = (column, task) => {
  boardsStore.selectedColumn = column;
  boardsStore.selectedCardTask = task;
  controllerStore.cardTaskView = true;
  controllerStore.overlay = true;
}
const handleDragStart = (task, columnIndex, taskIndex) => {
  const taskData = { task, columnIndex, taskIndex }
  boardsStore.setDraggedCardTask(taskData);
  console.log('Task Card data:', droppedTaskCard);
}
const handleOndrop = (columnIndex) => {
  boardsStore.drop(columnIndex);
}
const handleDragOver = (event) => {
  event.preventDefault();
}
</script>

<template>
  <div class="flex gap-6">
    <section  
      data-dragscroll 
      v-if="searchTerm" 
      v-for="(columnS, columnIndexS) in columns" :key="columnIndexS" 
      class="min-w-[280px] last:pr-6 box-content"
      @drop="handleOndrop(columnIndexS)"
      @dragover="handleDragOver"
    >
      <div class="flex items-center gap-3 pb-6 ">
        <div class="rounded-full h-4 w-4" :style="{ backgroundColor: bulletColors(columnIndexS) }"></div>
        <h2 class="text-medium-grey font-bold text-xs uppercase">
          {{ columnS.name }} ( {{ columnS.tasks.length }} )
        </h2>
      </div>
      <div
        v-for="(taskF, taskIndexF) in boardsStore.filteredTasks(columnS?.tasks)" 
        :key="taskIndexF"
        class="list-none cursor-grab active:cursor-grabbing active:animate-pulse"
        draggable="true"
        @dragstart="handleDragStart(taskF, columnIndexS, taskIndexF)"
      >
        <CardTask 
          @click="onClickTask(columnIndexS, taskIndexF)" 
          :task="taskF" 
          />
      </div>
    </section>
    <section 
      data-dragscroll 
      v-else 
      v-for="(column, columnIndex) in boardsStore.getColumns" 
      :key="columnIndex" 
      class="bg-lines-light min-w-[280px] last:pr-6"
      @drop="handleOndrop(columnIndex)"
      @dragover="handleDragOver"
    >        
      <div class="flex gap-3 pb-6">
        <div class="rounded-full h-4 w-4" :style="{ backgroundColor: bulletColors(columnIndex) }"></div>
        <h2 class="text-medium-grey font-bold text-xs uppercase">
          {{ column.name }} ( {{ column.tasks.length }} )
        </h2>
      </div>
        <div
          v-for="(task, taskIndex) in column?.tasks" 
          :key="taskIndex"
          class="list-none" 
          draggable="true"
          @dragstart="handleDragStart(task, columnIndex, taskIndex)"
        >
          <CardTask v-model="column.tasks" @click="onClickTask(columnIndex, taskIndex)" :task="task" />
    </div>
    </section>
  </div>
</template>


