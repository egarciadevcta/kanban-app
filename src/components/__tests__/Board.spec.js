import { describe, it, expect, vitest } from "vitest";
import { shallowMount } from "@vue/test-utils";
import { createTestingPinia } from "@pinia/testing";
import { useBoardStore } from "../../stores/board";
import Board from "../Board.vue";

describe("Board", () => {
  it("renders correctly", async () => {
    const wrapper = shallowMount(Board, {
      global: {
        plugins: [createTestingPinia({ createSpy: vitest.fn })],
      },
    });
    const store = useBoardStore;

    expect(wrapper.html()).toMatchSnapshot();
  });
});
