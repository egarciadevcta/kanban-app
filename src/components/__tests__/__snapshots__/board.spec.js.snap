// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Board > renders correctly 1`] = `"<div class="flex gap-6 items-start"></div>"`;
